ako-gateway-api/k8s/validator.go

```diff
@@ -166,8 +157,11 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
 	for _, gatewayInNamespace := range gatewayNsList {
 		if gateway.Name != gatewayInNamespace.Name {
 			for _, gwListener := range gatewayInNamespace.Spec.Listeners {
-				if *listener.Hostname == *gwListener.Hostname || isRegexMatch(string(*listener.Hostname), string(*gwListener.Hostname), key) || isRegexMatch(string(*gwListener.Hostname), string(*listener.Hostname), key) {
-					utils.AviLog.Errorf("key: %s, msg: Hostname overlaps or is same as an existing gateway %s hostname %s", key, gatewayInNamespace.Name, *gwListener.Hostname)
+				if gwListener.Hostname == nil {
+					continue
+				}
+				if listener.Hostname != nil && *listener.Hostname == *gwListener.Hostname {
```

Comment:
Also if we are planning to keep this check forever, due to limitations on controller side, can we use `gatewayToHostnameStore` here to list hostnames associated with a gateway instead of looping through gateway hostnames for consistency?

Any thoughts ?
@akshayhavile @aaha97

Reply:
@pkoshtavmware : This change has to be taken separately. Reasons:
1. `gatwewayToHostnameStore` stores hostnames associated with gateway. Currently, this is done during dequeueingestion phase which is always called after validation of gateway done with other gateway hostnames.
2. We can store these values in this data structure upfront in validation phase.
3. This will also require another datastructure `hostsToGateway` list to maintain hostname to Gateways.
4. So when any gateway gets deleted or listeners hostname changes, we need to fetch gateway associated with that hostname. Remove entry of deleted or changed gateway and then trigger add/update for other gateway.
5. So this will create of flow of deletion, addition of parents, children virtualservice. 
6. During bootup, we can arrange gateway in creation timestamp basis(which we do today also but after validaton) and then update this datastructure so no need to maintain anything at Controller side about gateway <-> hostname relations.